osu.Game.Tests/Visual/Gameplay/TestSceneMultiplayerGameplayLeaderboard.cs

```diff
@@ -0,0 +1,155 @@
+        private TestMultiplayerStreaming streamingClient = new TestMultiplayerStreaming(16);
+
+        [Cached(typeof(UserLookupCache))]
+        private UserLookupCache lookupCache = new TestSceneCurrentlyPlayingDisplay.TestUserLookupCache();
+
+        private MultiplayerGameplayLeaderboard leaderboard;
+
+        protected override Container<Drawable> Content { get; } = new Container { RelativeSizeAxes = Axes.Both };
+
+        public TestSceneMultiplayerGameplayLeaderboard()
+        {
+            base.Content.Children = new Drawable[]
+            {
+                streamingClient,
+                lookupCache,
+                Content
+            };
+        }
+
+        [SetUpSteps]
+        public void SetUpSteps()
+        {
+            AddStep("create leaderboard", () =>
+            {
+                OsuScoreProcessor scoreProcessor;
+                Beatmap.Value = CreateWorkingBeatmap(Ruleset.Value);
+
+                var playable = Beatmap.Value.GetPlayableBeatmap(Ruleset.Value);
+
+                streamingClient.Start(Beatmap.Value.BeatmapInfo.OnlineBeatmapID ?? 0);
+
+                Children = new Drawable[]
+                {
+                    scoreProcessor = new OsuScoreProcessor(),
+                };
+
+                scoreProcessor.ApplyBeatmap(playable);
+
+                LoadComponentAsync(leaderboard = new MultiplayerGameplayLeaderboard(scoreProcessor, streamingClient.PlayingUsers.ToArray())
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                }, Add);
+            });
+
+            AddUntilStep("wait for load", () => leaderboard.IsLoaded);
+        }
+
+        [Test]
+        public void TestScoreUpdates()
+        {
+            AddRepeatStep("update state", () => streamingClient.RandomlyUpdateState(), 100);
+        }
+
+        public class TestMultiplayerStreaming : SpectatorStreamingClient
+        {
+            public new BindableList<int> PlayingUsers => (BindableList<int>)base.PlayingUsers;
+
+            private readonly int totalUsers;
+
+            public TestMultiplayerStreaming(int totalUsers)
+            {
+                this.totalUsers = totalUsers;
+            }
+
+            public void Start(int beatmapId)
+            {
+                for (int i = 0; i < totalUsers; i++)
+                {
+                    ((ISpectatorClient)this).UserBeganPlaying(i, new SpectatorState
+                    {
+                        BeatmapID = beatmapId,
+                        RulesetID = 0,
+                    });
+                }
+            }
+
+            private readonly Dictionary<int, FrameHeader> lastHeaders = new Dictionary<int, FrameHeader>();
+
+            public void RandomlyUpdateState()
+            {
+                foreach (var userId in PlayingUsers)
+                {
+                    if (RNG.Next(0, 1) == 1)
+                        continue;
```

Comment:
This is actually always true right now, because `RNG.Next` is max-exclusive. I'm going to assume this meant to be a 50-50 split and change this to a `RNG.NextBool()` as I have a few other changes queued up anyway.

Reply:
i was wondering why something felt off with the test display!